### PR TITLE
Refs #30862 - do not validate new settings

### PR DESCRIPTION
**Pull request by ezr-ondrej**
While creating the settings, the definitions are not loaded yet, so the
validations depending on other setting values are failing. We should not
validate on create anyway.

This is fixing a forgotten validation on new setting from 8137b14c.

### CODE REVIEW

**Comment by ekohl** [comment]
The upgrade failure looks relevant:
```
PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_settings_on_name"
```

**Comment by ezr-ondrej** [comment]
Yeah it is... sorry for not sharing, I'm working on that one

**Comment by ezr-ondrej** [comment]
[test foreman]

**Comment by ezr-ondrej** [comment]
I've tested that this indeed fixes the nighly issue and is imho correct, we don't need to validate on create.

File: app/models/setting.rb [hunk 1]
```diff
@@ -49,7 +49,7 @@ def validate(record)
   validates :value, :format => { :with => Resolv::AddressRegex }, :if => proc { |s| IP_ATTRS.include? s.name }
   validates :value, :regexp => true, :if => proc { |s| REGEXP_ATTRS.include? s.name }
   validates :value, :array_type => true, :if => proc { |s| s.settings_type == "array" }
-  validates_with ValueValidator, :if => proc { |s| s.respond_to?("validate_#{s.name}") }
+  validates_with ValueValidator, if: proc { |s| s.respond_to?("validate_#{s.name}") }, on: :update
```
Comment (ekohl) [inline app/models/setting.rb]: Should the proc actually check if the definitions are loaded **and** it responds to the validate method?

**Review by ekohl — COMMENTED** [review]
Can you elaborate why the definitions aren't loaded on creation? Is that something we should fix instead?

**Comment by ezr-ondrej** [comment]
> Can you elaborate why the definitions aren't loaded on creation? Is that something we should fix instead?

That's by design as we are loading values right after the definitions. In a new concept, the defaults in database won't exist at all, so it is quite temporary issue.
But I can split the loading to two phases and load the definitions asap, although some definitions will not be present yet, as some definitions are loaded from the database, so we will need to reload anyway.

**Comment by ezr-ondrej** [comment]
> Is that something we should fix instead?

I think, this will help :+1:

File: app/services/setting_registry.rb [hunk 2]
```diff
@@ -44,6 +44,16 @@ def []=(name, value)
 
   def load
     # add() all setting definitions
+    load_definitions
+
+    # load all db values
+    load_values
+
+    # if create_missing create missing values in DB
+    # not needed now as old load mechanism takes care of that
```
Comment (ekohl) [inline app/services/setting_registry.rb:53]: This is a bit hard to parse for me. What is the old load mechanism?

**Review by ekohl — COMMENTED** [review]
This looks safer to me.

File: app/services/setting_registry.rb (hunk 2 above)
Comment (ezr-ondrej) [reply app/services/setting_registry.rb:53]: `Setting.descendants.load_defaults`

Comment (ezr-ondrej) [reply app/services/setting_registry.rb:53]: The way we do things now - create all the settings.
This is a note to keep in mind that if we would like to still have all the key/value pairs in db, this would be the place to create those :)

Comment (ekohl) [reply app/services/setting_registry.rb:53]: I think this would be better, but I still can't understand the `if` part.
```suggestion
    # if create_missing creates missing values in the DB
    # not needed now as the old load mechanism takes care of that
```

And I would not use `old load mechanism` but rather describe it. Or define a method `old_load_mechanism` and call that.

What I'm mostly trying to say is that this comment created more confusion on my side than that it helped.

Comment (ezr-ondrej) [reply app/services/setting_registry.rb:53]: Is it ok with you if I do this in the DSL PR, so we fix nightlies?

Comment (ekohl) [reply app/services/setting_registry.rb:53]: Sounds good.

**Review by ekohl — APPROVED** [review]
